File: cmd/secrt/create.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"

	"secret/internal/envelope"
)

func runCreate(args []string, deps Deps) int {
	pa, err := parseFlags(args, nil)
	if err == errShowHelp {
		printCreateHelp(deps)
		return 0
	}
	if err != nil {
		writeError(deps.Stderr, pa.json, err.Error())
		return 2
	}
	resolveGlobals(&pa, deps)

	// Read plaintext from exactly one source
	plaintext, err := readPlaintext(pa, deps)
	if err != nil {
		writeError(deps.Stderr, pa.json, err.Error())
		return 2
	}

	// Parse TTL
	var ttlSeconds *int64
	if pa.ttl != "" {
		ttl, err := envelope.ParseTTL(pa.ttl)
		if err != nil {
			writeError(deps.Stderr, pa.json, fmt.Sprintf("invalid TTL: %v", err))
			return 2
		}
		ttlSeconds = &ttl
	}

	// Resolve passphrase
	passphrase, err := resolvePassphraseForCreate(pa, deps)
	if err != nil {
		writeError(deps.Stderr, pa.json, err.Error())
		return 2
	}

	// Seal envelope
	result, err := envelope.Seal(envelope.SealParams{
		Plaintext:  plaintext,
		Passphrase: passphrase,
		Rand:       deps.Rand,
	})
	if err != nil {
		writeError(deps.Stderr, pa.json, fmt.Sprintf("encryption failed: %v", err))
		return 1
	}

	// Upload to server
	client := &APIClient{
		BaseURL:    pa.baseURL,
		APIKey:     pa.apiKey,
		HTTPClient: deps.HTTPClient,
	}

	resp, err := client.Create(CreateRequest{
		Envelope:   result.Envelope,
		ClaimHash:  result.ClaimHash,
		TTLSeconds: ttlSeconds,
	})
	if err != nil {
		writeError(deps.Stderr, pa.json, err.Error())
		return 1
	}

	// Output
	shareLink := envelope.FormatShareLink(resp.ShareURL, result.URLKey)

	if pa.json {
		out := map[string]interface{}{
			"id":         resp.ID,
			"share_url":  resp.ShareURL,
			"share_link": shareLink,
			"expires_at": resp.ExpiresAt,
		}
		enc := json.NewEncoder(deps.Stdout)
		enc.Encode(out)
	} else {
		fmt.Fprintln(deps.Stdout, shareLink)
	}

	return 0
}

func readPlaintext(pa parsedArgs, deps Deps) ([]byte, error) {
	sources := 0
	if pa.text != "" {
		sources++
	}
	if pa.file != "" {
		sources++
	}

	if sources > 1 {
		return nil, fmt.Errorf("specify exactly one input source (stdin, --text, or --file)")
	}

	if pa.text != "" {
		if pa.text == "" {
			return nil, fmt.Errorf("--text value must not be empty")
		}
		return []byte(pa.text), nil
	}

	if pa.file != "" {
		data, err := os.ReadFile(pa.file)
		if err != nil {
			return nil, fmt.Errorf("read file: %w", err)
		}
		if len(data) == 0 {
			return nil, fmt.Errorf("file is empty")
		}
		return data, nil
	}

	// stdin
	if deps.IsTTY() {
		fmt.Fprint(deps.Stderr, "Enter secret (Ctrl+D to finish):\n")
	}

	data, err := io.ReadAll(deps.Stdin)
	if err != nil {
		return nil, fmt.Errorf("read stdin: %w", err)
	}
	if len(data) == 0 {
		return nil, fmt.Errorf("input is empty")
	}
	return data, nil
}

File: src/create.rs
use std::fs;
use std::io::{Read, Write};

use crate::cli::{parse_flags, print_create_help, resolve_globals, CliError, Deps, ParsedArgs};
use crate::client::CreateRequest;
use crate::envelope::{self, format_share_link, SealParams};
use crate::passphrase::{resolve_passphrase_for_create, write_error};

pub fn run_create(args: &[String], deps: &mut Deps) -> i32 {
    let mut pa = match parse_flags(args) {
        Ok(pa) => pa,
        Err(CliError::ShowHelp) => {
            print_create_help(deps);
            return 0;
        }
        Err(CliError::Error(e)) => {
            write_error(&mut deps.stderr, false, &e);
            return 2;
        }
    };
    resolve_globals(&mut pa, deps);

    // Read plaintext from exactly one source
    let plaintext = match read_plaintext(&pa, deps) {
        Ok(p) => p,
        Err(e) => {
            write_error(&mut deps.stderr, pa.json, &e);
            return 2;
        }
    };

    // Parse TTL
    let ttl_seconds = if !pa.ttl.is_empty() {
        match envelope::parse_ttl(&pa.ttl) {
            Ok(ttl) => Some(ttl),
            Err(e) => {
                write_error(&mut deps.stderr, pa.json, &format!("invalid TTL: {}", e));
                return 2;
            }
        }
    } else {
        None
    };

    // Resolve passphrase
    let passphrase = match resolve_passphrase_for_create(&pa, deps) {
        Ok(p) => p,
        Err(e) => {
            write_error(&mut deps.stderr, pa.json, &e);
            return 2;
        }
    };

    // Seal envelope
    let result = envelope::seal(SealParams {
        plaintext,
        passphrase,
        rand_bytes: &*deps.rand_bytes,
        hint: None,
        iterations: 0,
    });

    let result = match result {
        Ok(r) => r,
        Err(e) => {
            write_error(
                &mut deps.stderr,
                pa.json,
                &format!("encryption failed: {}", e),
            );
            return 1;
        }
    };

    // Upload to server
    let client = (deps.make_api)(&pa.base_url, &pa.api_key);

    let resp = match client.create(CreateRequest {
        envelope: result.envelope,
        claim_hash: result.claim_hash,
        ttl_seconds,
    }) {
        Ok(r) => r,
        Err(e) => {
            write_error(&mut deps.stderr, pa.json, &e);
            return 1;
        }
    };

    // Output
    let share_link = format_share_link(&resp.share_url, &result.url_key);

    if pa.json {
        let out = serde_json::json!({
            "id": resp.id,
            "share_url": resp.share_url,
            "share_link": share_link,
            "expires_at": resp.expires_at,
        });
        let _ = writeln!(deps.stdout, "{}", serde_json::to_string(&out).unwrap());
    } else {
        let _ = writeln!(deps.stdout, "{}", share_link);
    }

    0
}

fn read_plaintext(pa: &ParsedArgs, deps: &mut Deps) -> Result<Vec<u8>, String> {
    let mut sources = 0;
    if !pa.text.is_empty() {
        sources += 1;
    }
    if !pa.file.is_empty() {
        sources += 1;
    }

    if sources > 1 {
        return Err("specify exactly one input source (stdin, --text, or --file)".into());
    }

    if !pa.text.is_empty() {
        return Ok(pa.text.as_bytes().to_vec());
    }

    if !pa.file.is_empty() {
        let data = fs::read(&pa.file).map_err(|e| format!("read file: {}", e))?;
        if data.is_empty() {
            return Err("file is empty".into());
        }
        return Ok(data);
    }

    // stdin
    if (deps.is_tty)() {
        let _ = writeln!(deps.stderr, "Enter secret (Ctrl+D to finish):");
    }

    let mut data = Vec::new();
    deps.stdin
        .read_to_end(&mut data)
        .map_err(|e| format!("read stdin: {}", e))?;
    if data.is_empty() {
        return Err("input is empty".into());
    }
    Ok(data)
}
